design-proposals/vm-console-logs.md

```diff
@@ -0,0 +1,743 @@
+# Overview
+Linux and Windows Server VMs can log boot and application data to their serial console.
+KubeVirt user can great benefit from having VMs logs, collected over the serial console, integrated into the Kubernetes logging architecture. 
+
+## Motivation
+VM logs streamed over the serial console contain really useful debug/troubleshooting info.
+While we are currently already capturing KubeVirt components logs (virt-controller, virt-launcher, libvirt, virt-handler, etc…) but the VM guest OS and application logs are not currently captured or indexed using cluster logging:
+currently KubeVirt lacks the level of integration that allows VM guests logs to be seamlessly streamed and captured as for generic container workloads ones.
+
+## Goals
+- It should let the user collect serial console logs for his VM according to the standard k8s logging architecture.
+- No guest OS sided changes should be required.
+- It should not prevent the usage of the serial console to interactively login into the VM.
+
+## Non Goals
+- VM serial console logs could be streamed to external logging stacks like Fluentd+Elasticsearch+Kibana/Loki/... but this specific part of the integration is out of scope for this proposal.
+- Defining custom/new format/structured for guest OS logs is out of scope. 
+
+## Definition of Users
+- VM owner: the user who owns a VM in his namespace on a Kubernetes cluster with KubeVirt
+- cluster-admin: the administrator of the cluster
+
+## User Stories
+- As a VM owner I want to view the VM guest OS logs using standard client tooling just as I would an application running in a container.
+- As a VM user I want to query my logs pertaining to an application running within my VM workload using Kubernetes standard logging
+- As a cluster-admin, I would like to view all the logs of all the VMs running in my cluster in order to find and correct reoccurring bugs, security flaws or misconfigurations.
+
+## Repos
+- https://github.com/kubevirt/kubevirt
+
+# Design
```

Comment:
How will this work with migration?

The example below focuses on the virt-launcher pod, but this is not a common KubeVirt user API.
A VM owner may not even have access to pods (I would be surprise if he had).
Can the VM object be the one the user interact with? If it can resolve the right pod in the background, it would be very nice.

Reply:
> Exactly as for libvirt logs that are streamed in the logs of the compute container

The guest logs focus on the guest while the libvirt (and virt-launcher) ones focus on the infrastructure that runs it. The fact that a migration translates into two pods is an implementation detail, same as libvirt and having a pod is an implementation detail.
For VM as a frontend to users, we usually provide an abstraction so they will not need to care about the backend details.

> If the user wants to get all the logs he has to setup a log collection and aggregation system as for a generic application and query it by VM name.

IMO, the minimum is to provide the generic instructions of how users can work with it (e.g. track the pods that are used). A better solution is to provide an abstraction and not burden a user with this, but I recognize this may be a follow up and harder to implement.

> The whole idea behind Kubevirt is to be able to have a VM natively interact with regular pods in an hybrid scenario so I'd expect an user to be able to start VMs and regular pods

I think there is a distinction between working on the same platform (VMs and Pods) and having access to these resources. Users that create VM/s may be given access to create and modify the VM CR, but not access to control the low level Pod. Accessing pods is a pretty privileged operation, as it allows one to run whatever they want on the platform/cluster.
Usually abstractions are given, limiting what can be run.

But you could claim that the user in question is the project admin (i.e. the namespace admin) which has the highest priv in the namespace, and has access to pods and SAs.

If the user is just the one that owns a specific VM, or has access only to VM abstractions (the VM CR), then accessing pods should not be an option for them.
On the other hand, if the user in question is the project admin, then this will make more sense.

As long as this is clear from the design point of view and expressed in the documentation, I think the limitations here are acceptable. If we will need to enhance the solution in the future, abstractions can be build over this base.

Reply:
> The guest logs focus on the guest while the libvirt (and virt-launcher) ones focus on the infrastructure that runs it. The fact that a migration translates into two pods is an implementation detail, same as libvirt and having a pod is an implementation detail.
For VM as a frontend to users, we usually provide an abstraction so they will not need to care about the backend details.

The main idea behind this is to be 100% tight to the [standard k8s logging architecture](https://kubernetes.io/docs/concepts/cluster-administration/logging/) to get the expected integration with all the logs aggregating solutions (like the ELK stack, Loki, ...) for free.

> IMO, the minimum is to provide the generic instructions of how users can work with it (e.g. track the pods that are used). A better solution is to provide an abstraction and not burden a user with this, but I recognize this may be a follow up and harder to implement.

virt-launcher pods are already labeled with something like:
```yaml
  labels:
    kubevirt.io: virt-launcher
    kubevirt.io/domain: rhel9-above-urial
    vm.kubevirt.io/name: rhel9-above-urial
```
this is only about using the right labels when querying the log management tool.

Reply:
TBH I think I agree with @EdDev's point here.

Kubevirt's premise is that users would be able to run VMs alongside "regular" pods. However, the fact that a VM runs inside a pod is indeed an implementation detail. The user doesn't have to know or care about the virt-launcher pod itself and as of today, there's no reason to interact with this pod to inspect the guest itself, but only in order to inspect the virtualization stack that's running the guest. All of the user's interaction with the VM is done by the VirtualMachine CRD which is the user-facing k8s object.

However, implementation-wise I'm not sure if it's possible / easy to redirect the logs to the VM object itself. Is it common to perform operations like `kubectl logs vm my-vm` on other CRDs? How are other projects coping with this?

Perhaps as a compromise we can hide the implementation from the user by introducing a `virtctl log my-vm` command that would know to fetch this logs without complicating the users with the need to understand that the VM runs inside a pod, and that there's a side-car container that actually fetches these logs, etc.

WDYT?

Reply:
> However, implementation-wise I'm not sure if it's possible / easy to redirect the logs to the VM object itself. Is it common to perform operations like `kubectl logs vm my-vm` on other CRDs? How are other projects coping with this?

As for:
https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#advanced-features-and-flexibility
`Other Subresources - Add operations other than CRUD, such as "logs" or "exec".` is not supported on CRDs.

> Perhaps as a compromise we can hide the implementation from the user by introducing a `virtctl log my-vm` command that would know to fetch this logs without complicating the users with the need to understand that the VM runs inside a pod, and that there's a side-car container that actually fetches these logs, etc.
> 
> WDYT?

Introducing `virtctl logs -n <vmnamespace> <vmname>` as wrapper for `kubectl logs -n <vmnamespace> -l vm.kubevirt.io/name=<vmname>` is probably straightforward but I want also to mention that interactively fetching the logs of a running VM is just one of the possible use-cases. 
I'd expect that being able to stream the guest console of the VMs on the cluster to an external logging stack (like [Grafana Loki](https://grafana.com/oss/loki/)) to store, analyze, and query it there it will become pretty common.
All the k8s logging stack are designed to natively manage and filter pod logs according to the standard k8s logging architecture, if we design something significantly different it will risk to require further effort to be integrated into other tools.
Currently with this approach we can easily get something like this in Loki basically out of the box:
[guest console log in Grafana Loki](https://pasteboard.co/khxx1VDKhQUN.png)